### ISSUE TITLE
Github Desktop unexpectedly closes on version 2.6.4

### ISSUE THREAD

**Issue by sky93** [desktop/desktop#11696]
### Describe the bug

After I updated my Github Desktop from version 2.6.3 to 2.6.4 when I open the app and click on Fetch origin, App unexpectedly closes. I don't have any issues with version 2.6.3. It just happened from version 2.6.4 and I cannot use Github Desktop at all. 

By the way, after Github Desktop closes I still see 2-3 Github Desktio process in Task Manager. I have to kill them and launch the app again, Otherwise, it won't open.

### Version & OS

Github Desktop: 2.6.4
OS: Windows 10

### Steps to reproduce the behavior

1. Just click on the Fetch origin button

### Expected behavior

Fetch from origin.

### Actual behavior

Application closes.

### Logs

```
2021-03-04T12:42:32.904Z - info: [ui] [AppStore] loading 189 repositories from store
2021-03-04T12:42:32.905Z - info: [ui] [AppStore] found account: sky93 (Sepehr Mohaghegh)
2021-03-04T12:42:33.195Z - info: [ui] launching: 2.6.4 (Windows 10.0.19041)
2021-03-04T12:42:33.199Z - info: [ui] execPath: 'C:\Users\Sky\AppData\Local\GitHubDesktop\app-2.6.4\GitHubDesktop.exe'
2021-03-04T12:42:37.413Z - error: [main] Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:205:27)
2021-03-04T12:42:38.124Z - info: [main] Error report submitted
```

It seems I have to revert to version 2.6.3.

**Comment by billygriffin** [comment]
Hi @sky93, thanks so much for reporting and sorry you're dealing with this. We're working on a fix right now at least for the app crash, and trying to figure out what might be the underlying cause. Would you mind sharing if there's anything notable about your environment? Like a proxy, antivirus, or anything else that might help us diagnose the root cause? Thanks!

**Comment by sky93** [comment]
Thank you @billygriffin, No, I don't have any special proxies or antivirus. Just using Windows Defender as my antivirus. When I open Github Desktop again I see that it fetched the origin successfully but it closes.

**Comment by billygriffin** [comment]
Thanks so much @sky93! A fix for this is being deployed right now to production. After that ships, we're also going to ship something that displays and error with more information instead of crashing to our beta channel. Given that this is an exceptionally rare crash, would you be willing to try out the beta when that ships and share what the error is to help us diagnose? We'll share that here again when it's available (and when the fix is available on prod).

**Comment by sky93** [comment]
Yes I'd love too!

**Comment by sky93** [comment]
I can confirm that this issue is fixed in version 2.6.5.

 Thank you.

**Comment by sergiou87** [comment]
@sky93 Thank you for confirming that fix works! ❤️  As @billygriffin suggested, could you also try the latest beta (v2.6.6-beta1)? You can download it from here: https://github.com/desktop/desktop/#beta-channel

That beta will display an error message to you where the previous production build used to crash. It would be very helpful if you could try to reproduce those errors a few times and share the error messages here.

**Comment by sky93** [comment]
Yes, I switched to the beta channel and now I can see the error message when fetching from the origin:

![image](https://user-images.githubusercontent.com/8404511/110010996-2821b080-7d34-11eb-9f6a-ab75375b608f.png)

```
ERROR: Couldn't connect to 127.0.0.1:52269: 10061
error: unable to read askpass response from 'C:\Users\Sky\AppData\Local\GitHubDesktop\app-2.6.6-beta1\resources\app\desktop-trampoline\desktop-trampoline.exe'
fatal: could not read Password for 'https://sky93@git.example.com': terminal prompts disabled
```
In the screenshot above my repository is from a Gitlab CE server. After this test, I tried to clone one of my private repositories from Github but I couldn't:
![image](https://user-images.githubusercontent.com/8404511/110011858-21e00400-7d35-11eb-8daa-53989b14571f.png)

```
Cloning into 'D:\sample'...
ERROR: Couldn't connect to 127.0.0.1:52977: 10061
error: unable to read askpass response from 'C:\Users\Sky\AppData\Local\GitHubDesktop\app-2.6.6-beta1\resources\app\desktop-trampoline\desktop-trampoline.exe'
fatal: could not read Password for 'https://sky93@github.com': terminal prompts disabled

Would you like to retry cloning?
```

After that, I tried to clone a public repo from Github and it cloned without any errors. So I guess there's something wrong with reading cached passwords when working with private repos?

**Comment by sergiou87** [comment]
Thank you @sky93 that's super useful!!

The line I was looking for is `ERROR: Couldn't connect to 127.0.0.1:52977: 10061` and to be honest I'm very surprised to find that 😂 

With this I have enough to start digging, thank you for taking the time!! ❤️

**Comment by sergiou87** [comment]
I haven't been able to reproduce that error on my end @sky93, but I was wondering if there is anything specific about your network configuration that could cause this issue, for example:
1. Do you have any kind of firewall rule that could affect incoming TCP connections to 127.0.0.1? If you have a Firewall at all, could you try disabling it entirely and then testing the beta to see if that solves the issue?
1. Does your machine have anything special around 127.0.0.1? Is your machine/network configured for IPv6 only by any chance?
1. Do you have another machine where you can test the beta build and see if the problem occurs there too?

So far I think the most plausible option is the existence of a firewall blocking the connection. If you use Windows Defender, you're probably using Windows Firewall too, right?

**Comment by sky93** [comment]
No, I'm not using IPv6. I disabled my Windows Defender and firewall completely and I still have this issue. 

By the way, when I try to clone a repo in command line, a prompt pops up and when I enter my **correct** username and password it says Login failed and enter credentials using the basic prompt: 

![image](https://user-images.githubusercontent.com/8404511/110159722-4fde4a80-7e00-11eb-998d-1f9db7dce38e.png)

![image](https://user-images.githubusercontent.com/8404511/110160803-bc0d7e00-7e01-11eb-9577-1aca70a0b513.png)

I really don't know if this relates to this issue.

And I checked this on another machine too and I had no problem. It's probably related to my machine only but I don't think I'm the only one as you see in #11701. If there's anything else that I can test please let me know.

**Comment by steveward** [comment]
@sky93 that authentication dialog is standard for Git for Windows. You need to [generate a token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) to authenticate there.

Can you try running `ping 127.0.0.1` from the command prompt to see what returns?

**Comment by sky93** [comment]
> @sky93 that authentication dialog is standard for Git for Windows. You need to [generate a token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) to authenticate there.
> 
> Can you try running `ping 127.0.0.1` from the command prompt to see what returns?

Well, it seems normal.

![image](https://user-images.githubusercontent.com/8404511/110177283-fbe05f80-7e19-11eb-8aad-0b9b5002727a.png)

**Comment by sergiou87** [comment]
Thanks for checking that, @sky93, we really appreciate it!

For another user (https://github.com/desktop/desktop/issues/11701#issuecomment-792454746) the problem was the antivirus/firewall. I'm a bit surprised that doesn't fix it for you too 😞 

Do you have another Windows computer where you could try that beta out? I wonder if it's something specific to your machine, or if it's something else in your environment (not sure how, but maybe network related?).

I might have a new test build later today with some extra checks to help diagnosing the issue, if you could try it out.

Thanks again for your help! 🙇‍♂️

**Comment by sky93** [comment]
I checked another Windows machine and it was fine. Yeah, the problem is with my machine even though I completely disabled my antivirus. And yes I'd be happy to try it out if it's helpful to you.

**Comment by sergiou87** [comment]
Hi again @sky93! Sorry for the delay, I didn't have time to work on this yesterday. This is the link for the test build I mentioned: https://s3.amazonaws.com/github-desktop/releases/2.6.6-test1-ca3234d4/GitHubDesktopSetup.exe

Now what I'd need is that you check in the Console of the Developer Tools (you can open them from the `View` menu, then `Toggle Developer Tools`) and see if there are messages containing `internal trampoline test` when you perform any operation like cloning, fetching, pulling or pushing.

Something like this:

https://user-images.githubusercontent.com/1083228/110472175-54934f00-80dd-11eb-8ddf-e953cdc0d844.mov

You should see either a `Internal trampoline test to port ${port} succeeded?` with `true` or `false`, or `Internal trampoline test to port ${port} failed` and then the exact error.

**Comment by sky93** [comment]
![image](https://user-images.githubusercontent.com/8404511/110716923-5fed9400-821d-11eb-830b-54adfbdb1074.png)

This is my result. It looks strange to me as your Internal trampoline test is succeeded but I get the error.

**Comment by sergiou87** [comment]
Thanks for trying that build out @sky93, that's super helpful!

The result is definitely weird 😐 

The internal test is connecting to port 52537, however it's doing that from within the app. My bet again is that there is _something_ in your machine (like an antivirus or a firewall) preventing other binaries different than the app to connect to the port.

However you say you only have Windows Defender and its firewall and you have disabled both but it doesn't change anything, then… 🤯 

Something else you could try is running the Command Line **as an Administrator** (very important), and then run `netstat -qb` to see if the port is open.

Basically you need to follow the same steps you did to get to that screenshot, check the port number used, and then look for it in `netstat -qb | findstr <port number>` like this:

![image](https://user-images.githubusercontent.com/1083228/110774238-dca06280-825d-11eb-94f8-190c93fb46aa.png)

In theory you should see the same `LISTENING` state for that port

**Comment by sky93** [comment]
I tested it with `netstat`. It seems Github Desktop doesn't open a port so obviously trampoline cannot connect to it.

![image](https://user-images.githubusercontent.com/8404511/110785143-d6bb7900-827f-11eb-9af8-8b9af5adc74e.png)

![image](https://user-images.githubusercontent.com/8404511/110785208-ea66df80-827f-11eb-9773-d08e3549acc6.png)

I don't know what's trampoline but I tried to test if it's working. So I opened a port in netcat in WSL, created an environment variable called `DESKTOP_PORT` and opened up trampline and it was working correctly! Look at the video I just recorded:


https://user-images.githubusercontent.com/8404511/110786996-11beac00-8282-11eb-9a45-9419d9309034.mp4

**Comment by sergiou87** [comment]
This is great, thanks @sky93 !!! I think you found a different issue: when the trampoline client fails to connect (as it happens to you), we immediately kill the server 🤦 

I'm gonna fix that and make sure you can check if the port is listening, I'll post a link with the new test build ASAP. Other than that, what you show in the video is awesome, those tools could come very handy to debug this issue. Thank you!! 🙇‍♂️🙇‍♂️🙇‍♂️🙇‍♂️

**Comment by sky93** [comment]
I'm glad. Hope you can find the main cause of this issue.

**Comment by sergiou87** [comment]
@sky93 the new test build: https://s3.amazonaws.com/github-desktop/releases/2.6.7-test1-38e89bb7/GitHubDesktopSetup.exe

This one should keep the port open after getting the "couldn't connect error", so you can `netstat` it 🤞

**Comment by sky93** [comment]
Good job! The problem is resolved now! So it wasn't a firewall issue after all and if the problem was closing the server on client error, I'm surprising why this happened to my machine only? Thank you so much @sergiou87 !

![image](https://user-images.githubusercontent.com/8404511/110805249-26a53a80-8296-11eb-8e30-f5046efeb00d.png)

**Comment by sergiou87** [comment]
OMG!!! Well this is crazy and unexpected 😂 

I'd expect that you still see the error, because that hasn't changed!

This is what happened in previous versions:
1. GH Desktop starts listening at some port.
2. Trampoline tries to connect to that port and fails. **This is when you see the error.**
3. GH Desktop catches the error and kills the server (🤦 )

The new version only changes (3), but (2) should still happen, no?? 🤔

**Comment by donbode** [comment]
I am an amateur.  17 min after a successful commit, I evoked another commit.  I got the error 
![image](https://user-images.githubusercontent.com/57908896/110815041-47926f00-8247-11eb-82f5-20fde3682700.png)
I looked in the recommended location and there was nothing beyond   \AppData\Local\GitHubDesktop\app-2.6.6\resources\app 
The best I can do with that address is 
C:\Users\boded\AppData\Local\GitHubDesktop\app-2.6.6\resources\app 
Is this related to the above problem?  What should I do.
Don

**Comment by donbode** [comment]
At that address is
![image](https://user-images.githubusercontent.com/57908896/110815700-e7e89380-8247-11eb-8523-1a798b3806c0.png)

**Comment by sergiou87** [comment]
Hi @donbode! 👋 

Looking at the message, sounds like you have some kind of antivirus that is deleting or putting git.exe and other binaries in quarantine. You can probably disable the antivirus or mark those binaries as "safe", somehow, that will depend on your antivirus 😅 

Other than that, I don't think this is the best place to ask. Next time, I suggest you use the green `New Issue` button at the top right:
![image](https://user-images.githubusercontent.com/1083228/110816226-716d8500-828b-11eb-8600-559ef9c6e419.png)

That way more people can help you more efficiently!

In any case, I hope my message helped 😄

**Comment by sky93** [comment]
> OMG!!! Well this is crazy and unexpected 😂
> 
> I'd expect that you still see the error, because that hasn't changed!
> 
> This is what happened in previous versions:
> 
> 1. GH Desktop starts listening at some port.
> 2. Trampoline tries to connect to that port and fails. **This is when you see the error.**
> 3. GH Desktop catches the error and kills the server (🤦 )
> 
> The new version only changes (3), but (2) should still happen, no?? 🤔

Well, I'm pretty sure the new version fixed the issue somehow! For making sure about this I downgraded to the previous test version you provided (https://github.com/desktop/desktop/issues/11696#issuecomment-793839021) and the error popped up again and installed the new version and the error is gone!

**Comment by sergiou87** [comment]
@sky93 I have another theory, not sure how crazy it is. If this last change actually fixed the problem you had, that could mean there is _something_ else in your machine connecting to this socket. This would be the timeline of events:

1. App starts the server at port XXXX
2. _Something_ in your machine connects to the server at port XXXX
3. That _something_ in your machine sends a [RST packet](https://www.pico.net/kb/what-is-a-tcp-reset-rst) and kills its connection (this is where the `ECONNRESET` from your original post would come from).
4. App catches that error and kills the server (**this is what I fixed in the last build**).
5. Trampoline tries to connect to the server at port XXXX, but it was closed in the previous step, so the connection fails and you see the error you originally reported.

Since my latest build fixed (4), in step (5) the trampoline connects successfully and everything works. But according to my theory there is still _something_ in your machine connecting to the server and sending that [RST packet](https://www.pico.net/kb/what-is-a-tcp-reset-rst).

One way to test this theory is to start your computer in safe mode, and try one of the older builds to see if it works. **If that _something_ I mentioned starts on login but not in safe mode** (which might not be the case), using safe mode should prevent that, and therefore older builds should work.

**Comment by sky93** [comment]
No problem, I'm gonna test it in safe mode tonight or tomorrow morning.

**Comment by sky93** [comment]
Just tested in safe mode. I confirm that both test versions working fine in safe mode. So your theory is correct :) I don't know what's wrong with my machine that kills the port Github Desktop listens to but as you don't kill the server on error my issue is resolved. Thank you @sergiou87 !

**Comment by sergiou87** [comment]
Yay!! Thanks for testing that!! And I'm super happy it works. Thank you so much for your help, it was key in finding the fix 🙇‍♂️ 🙇‍♂️ 🙇‍♂️ 🙇‍♂️
